File: Problems/Leetcode/3001-4000/3201-3300/3234.cpp
#include <bits/stdc++.h>

using namespace std;

static constexpr int MOD = 1e9 + 7;

class Solution {
public:
    int numberOfSubstrings(string s) {
        int n = s.size();

        // 记录 0 的坐标
        vector<int> ids;
        for (int i = 0; i < n; ++i) {
            if (s[i] == '0') {
                ids.emplace_back(i);
            }
        }

        int cnt = n - ids.size(); // 统计 1 的个数

        ids.emplace_back(n); // 哨兵

        int ans = 0;
        int start = 0;
        for (int i = 0; i < n; ++i) { // 枚举左端点
            for (int k = start; k < ids.size() - 1; ++k) { // 枚举左端点后续的 0
                int c0 = k - start + 1;
                int c1 = ids[k] - i + 1 - c0;
                if (c0 * c0 > cnt) {
                    break;
                }
                // 去掉一部分不合法的 1
                ans += max(ids[k + 1] - ids[k] - max(c0 * c0 - c1, 0), 0);
            }
            if (s[i] == '0') {
                start++;
            } else {
                ans += ids[start] - i; // 0 之前（全为 1 的情况）
                cnt--;
            }
        }
        return ans;
    }
};

int main() {
    Solution sol;

    return 0;
}

File: Problems/Leetcode/3001-4000/3201-3300/3234.py
from typing import List

fmax = lambda x, y: x if x > y else y
MOD = 1_000_000_007


class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        ids = [i for i, ch in enumerate(s) if ch == "0"]
        tot = n - len(ids)  # 1 的个数
        ids.append(n)  # 哨兵

        ans = 0
        start = 0  # 左端点
        for i, ch in enumerate(s):
            for k in range(start, len(ids) - 1):  # 枚举左端点后续的 0
                c0 = k - start + 1
                if c0 * c0 > tot:
                    break
                c1 = ids[k] - i + 1 - c0
                ans += fmax(ids[k + 1] - ids[k] - fmax(c0 * c0 - c1, 0), 0)

            if ch == "0":
                start += 1
            else:
                ans += ids[start] - i
                tot -= 1
        return ans
